provider/gateway/router.go

```diff
@@ -116,9 +148,121 @@ func leaseServiceStatusHandler(log log.Logger, cclient cluster.ReadClient) http.
 }
 
 func leaseServiceLogsHandler(log log.Logger, cclient cluster.ReadClient) http.HandlerFunc {
-	return func(w http.ResponseWriter, req *http.Request) {
-		http.Error(w, "unimplemented", http.StatusNotImplemented)
+	return func(w http.ResponseWriter, r *http.Request) {
+		upgrader := websocket.Upgrader{
+			ReadBufferSize:  1024,
+			WriteBufferSize: 1024,
+		}
+
+		ws, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			if _, ok := err.(websocket.HandshakeError); !ok {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+			return
+		}
+
+		wsLogWriter(r.Context(), ws, wsLogsConfig{
+			lid:       requestLeaseID(r),
+			service:   requestService(r),
+			follow:    requestLogFollow(r),
+			tailLines: requestLogTailLines(r),
+			log:       log,
+			client:    cclient,
+		})
+	}
+}
+
+func wsLogWriter(ctx context.Context, ws *websocket.Conn, cfg wsLogsConfig) {
+	pingTicker := time.NewTicker(pingPeriod)
+
+	cctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		pingTicker.Stop()
+		cancel()
+		_ = ws.Close()
+	}()
+
+	logs, err := cfg.client.ServiceLogs(cctx, cfg.lid, cfg.service, cfg.follow, cfg.tailLines)
+	if err != nil {
+		cfg.log.Error("couldn't fetch logs: %s", err.Error())
+		err = ws.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocketInternalServerErrorCode, err.Error()))
+		if err != nil {
+			cfg.log.Error("couldn't push control message through websocket: %s", err.Error())
+		}
+		return
+	}
+
+	if len(logs) == 0 {
+		_ = ws.WriteMessage(
+			websocket.CloseMessage,
+			websocket.FormatCloseMessage(websocketInternalServerErrorCode, "service "+cfg.service+" does not have running pods"))
+		return
 	}
+
+	_ = ws.SetReadDeadline(time.Now().Add(pongWait))
+	ws.SetPongHandler(func(string) error {
+		_ = ws.SetReadDeadline(time.Now().Add(pongWait))
+		return nil
+	})
+
+	go func() {
+		for {
+			if _, _, e := ws.ReadMessage(); e != nil {
+				if _, ok := e.(*websocket.CloseError); !ok {
+					_ = ws.Close()
+				}
+				break
+			}
+		}
+	}()
+
+	var scanners sync.WaitGroup
+
+	logCh := make(chan cluster.ServiceLogMessage)
+
+	scanners.Add(len(logs))
+
+	for _, lg := range logs {
+		go func(name string, scan *bufio.Scanner) {
+			defer scanners.Done()
+
+			for scan.Scan() {
+				logCh <- cluster.ServiceLogMessage{
+					Name:    name,
+					Message: scan.Text(),
+				}
+			}
+		}(lg.Name, lg.Scanner)
+	}
+
+	alive := true
+
+	for alive {
+		select {
+		case line := <-logCh:
+			if err = ws.WriteJSON(line); err != nil {
+				alive = false
+			}
+		case <-pingTicker.C:
+			if err = ws.WriteMessage(websocket.PingMessage, []byte{}); err != nil {
+				alive = false
+			}
+		}
+	}
+
+	for i := range logs {
+		_ = logs[i].Stream.Close()
+	}
+
+	// drain logs channel in separate goroutine to unblock seeders waiting for write space
+	go func() {
+		for range logCh {
+		}
+	}()
+
+	close(logCh)
+	scanners.Wait()
```

Comment:
There are still problems here.

1. There's no guarantee that all of the writes to `logCh` are done before it's closed.
1. There's no way to know that the scanners have finished in the `for alive {` loop.

Suggestion from [earlier](https://github.com/ovrclk/akash/pull/721#discussion_r448813951):

* worker: `scan.Close()` in the before `scanner.Done()` so you don't have to deal with it elsewhere.
* worker: guard the loop with `scan.Scan() && ctx.Err() == nil` so that it stops when `cancel()` is called.
* mainloop: kick off a go routine that does `scanner.Wait()` then can signal that to `select` by closing a channel.
* mainloop: call `cancel()` to begin shutdown after `for alive {` loop
* mainloop: properly drain `logCh` by selecting in a loop along with the channel that signals that the workers are done.  this should not be in a separate thread.

Reply:
right `scanners.Wait` should be this way to prevent write into closed channel
```go
scanners.Wait()
close(logCh)
```

- scanner does not implement `Close()` that why have to close it from the reader. 
- `Scan()` can be blocked by underlying `Read()` call. to properly abort the call `Close()` should be called from another goroutine

```
	for _, lg := range logs {
		go func(name string, scan *bufio.Scanner) {
			defer scanners.Done()

			for scan.Scan() {
				logCh <- cluster.ServiceLogMessage{
					Name:    name,
					Message: scan.Text(),
				}
			}
		}(lg.Name, lg.Scanner)
	}

	alive := true

	donech := make(chan struct{})

	go func() {
		scanners.Wait()
		close(donech)
	}()

	for alive {
		select {
		case line := <-logCh:
			if err = ws.WriteJSON(line); err != nil {
				alive = false
			}
		case <-pingTicker.C:
			if err = ws.WriteMessage(websocket.PingMessage, []byte{}); err != nil {
				alive = false
			}
		case <-donech:
			alive = false
		}
	}

	cancel()

	for i := range logs {
		_ = logs[i].Stream.Close()
	}

	// drain logs channel in separate goroutine to unblock seeders waiting for write space
	go func() {
		for range logCh {
		}
	}()

	close(logCh)
```

Reply:
no, example above is wrong. it will deadlock

Reply:
looks like original approach was right. `for alive {` does not need to know about streams state. any first failure from `ws.WriteJSON(line)` or writing ping message, or timeout on waiting for `pong` break the `alive` loop.
then all streams are forced to close
```go
	for i := range logs {
		_ = logs[i].Stream.Close()
	}
```

by draining `logch` we ensure all scanners waiting on space in `logch` write and exit. when `logch` is closed all scanners finished

Reply:
> does not need to know about streams state. any first failure from ws.WriteJSON(line) ...

I'm talking about the logs coming from kube - `scan.Scan()` having returned false for all of them.

> by draining logch we ensure all scanners waiting on space in logch write and exit. when logch is closed all scanners finished

When do you know that you should close `logCh`?  Why do you need to close `logCh`?

Reply:
sorted in slack